### PR TITLE
Fix sneaking (fixes #665 and #3045)

### LINKED ISSUE #329: Possible to avoid fall damage by sneaking
When you drop down from any heigh and land on a slab with shift key pressed you won't take damage.
This also occurs when landing on the lower step of a stair, on the upper step you will take normal damage.
I think it may be for all nodeboxes but i think it will work with all nodeboxes. This may be related to this bug:
https://github.com/celeron55/minetest/issues/127

### LINKED ISSUE #665: Sneaking thinks nodeboxes are full nodes
If you sneak on a nodebox, you will collide with the node like if it was a full node.

### PR DESCRIPTION
**Pull request by BlockMen**
Currently #321 is not fixed, #665 and #3045 is.

### CODE REVIEW

**Comment by est31** [comment]
This will cause in lots and lots of calls to the map, and copy the collision boxes every time. Perhaps better cache the collision boxes, and only change them if `m_sneak_node` changed, like [here](https://github.com/BlockMen/minetest/blob/fix_sneaking/src/localplayer.cpp#L312)?

**Comment by est31** [comment]
Also, I don't see how this fixes 665 completely, its only a partial fix, but honestly IMO 665 is not a bug.

If we decide to go with this partial fix, we don't even have to cache the collision boxes, only the max_y value is enough.

**Comment by BlockMen** [comment]
@est, yes you are right. The max_y is stored now and updated within m_sneak_node update.

Also it works for stairs now, so its fixing #665. And #665 is for sure a bug since it would still behave as full node currently, even if the collisionbox is defined different.

**Comment by est31** [comment]
nice, will look at it later.

File: src/localplayer.cpp [hunk 1]
```diff
@@ -240,17 +241,17 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 		m_need_to_get_new_sneak_node = false;
 		m_sneak_node_exists = false;
 	}
-	else if(nodemgr->get(map->getNodeNoEx(current_node)).name != "air")
+	else if (nodemgr->get(map->getNodeNoEx(current_node)).name != "air")
```
Comment (est31) [inline src/localplayer.cpp]: The braces should probably be on the same line

File: src/localplayer.cpp [hunk 2]
```diff
@@ -298,11 +299,24 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
 
 		m_sneak_node = new_sneak_node;
 		m_sneak_node_exists = sneak_node_found;
 
+		if (sneak_node_found) {
+			f32 cb_max = 0;
+			MapNode n = map->getNodeNoEx(m_sneak_node);
+			std::vector<aabb3f> nodeboxes = n.getCollisionBoxes(nodemgr);
+			for (std::vector<aabb3f>::iterator i = nodeboxes.begin();
+					i != nodeboxes.end(); i++) {
```
Comment (est31) [inline src/localplayer.cpp]: Can you pre-increment the iterator? post-increment will just create a needless  copy.

**Comment by kaeza** [comment]
This does not seem to fix the issue on my side.

**Comment by BlockMen** [comment]
@kaeza which issue exactly? #665? And at all, just stairs or other nodeboxes?

**Comment by kaeza** [comment]
Eh ignore that comment. It works (mostly):

* Walking up stairs/slabs works, and the player is not lifted the whole node.
* Walking down stairs/slabs does not work.

Definitely much better.

**Comment by BlockMen** [comment]
Walking down while sneaking would need a custom "search radius" for nodeboxes that would have to be a lot smaller. And for some such a change would work (e.g. stairs) but other would suffer (e.g. slabs) since you wouldn't be able to sneak-build like now.

**Comment by est31** [comment]
I don't think you should be abled to walk down nodeboxes while sneaking.

File: src/localplayer.cpp [hunk 3]
```diff
@@ -230,27 +228,28 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
-	if(m_sneak_node_exists &&
+	f32 position_y_mod = 0.05 * BS;
+	if (m_sneak_node_bb_ymax > 0)
+		position_y_mod = m_sneak_node_bb_ymax - position_y_mod;
+	v3s16 current_node = floatToInt(position - v3f(0, position_y_mod, 0), BS);
+	if (m_sneak_node_exists &&
 	   nodemgr->get(map->getNodeNoEx(m_old_node_below)).name == "air" &&
-	   m_old_node_below_type != "air")
-	{
+	   m_old_node_below_type != "air") {
```
Comment (est31) [inline src/localplayer.cpp]: Perhaps also make the indentation with tabs?

**Comment by est31** [comment]
I think what @kaeza meant is that if you fall onto e.g. a slab, you still stop at the full node border, and not the slab's actual height.

**Comment by BlockMen** [comment]
Marking as WIP until either #321 is removed from commit msg or gets fixed again

File: src/localplayer.cpp [hunk 4]
```diff
@@ -298,11 +297,24 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
 		}
 
-		bool sneak_node_found = (min_distance_f < 100000.0*BS*0.9);
+		bool sneak_node_found = (min_distance_f < 100000.0 * BS * 0.9);
 
 		m_sneak_node = new_sneak_node;
 		m_sneak_node_exists = sneak_node_found;
 
+		if (sneak_node_found) {
+			f32 cb_max = 0;
```
Comment (kwolekr) [inline src/localplayer.cpp]: Could this be precalculated?

Comment (kwolekr) [inline src/localplayer.cpp]: Hmm, nevermind this, I think we could optimize it later on.  The code here gets executed only when moving to a new node while sneaking.

File: src/localplayer.cpp [hunk 5]
```diff
@@ -179,25 +180,22 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+		if (!is_climbing) {
+			// Move up if necessary
+			f32 new_y = (lwn_f.Y - 0.5 * BS) + m_sneak_node_bb_ymax;
+			if (position.Y < new_y)
+				position.Y = new_y;
 
-				if(m_speed.Y < 0)
-					m_speed.Y = 0;
-			}
+			if (m_speed.Y < 0)
```
Comment (kwolekr) [inline src/localplayer.cpp]: This block now gets executed even when position.Y >= min_y.  Was this intentional?

**Pull request merged**